kvbc/src/categorization/shared_kv_category.cpp

```diff
@@ -0,0 +1,142 @@
+}
+
+// root_hash = h((h(k1) || h(v1)) || (h(k2) || h(v2)) || ... || (h(k3) || h(v3)))
+void updateCategoryHash(const std::string &category_id,
+                        const Hash &key_hash,
+                        const Hash &value_hash,
+                        std::map<std::string, Hasher> &category_hashers) {
+  auto [it, inserted] = category_hashers.emplace(category_id, Hasher{});
+  if (inserted) {
+    it->second.init();
+  }
+  it->second.update(key_hash.data(), key_hash.size());
+  it->second.update(value_hash.data(), value_hash.size());
+}
+
+std::string dataCf(const std::string &category_id) { return category_id + SHARED_KV_DATA_CF_SUFFIX; }
+
+std::string latestVersionCf(const std::string &category_id) { return category_id + SHARED_KV_LATEST_KEY_VER_CF_SUFFIX; }
+
+SharedKeyValueCategory::SharedKeyValueCategory(const std::shared_ptr<storage::rocksdb::NativeClient> &db) : db_{db} {}
+
+SharedKeyValueUpdatesInfo SharedKeyValueCategory::add(BlockId block_id,
+                                                      SharedKeyValueUpdatesData &&update,
+                                                      storage::rocksdb::WriteBatch &batch) {
+  if (update.kv.empty()) {
+    throw std::invalid_argument{"Empty shared key-value updates"};
+  }
+
+  auto update_info = SharedKeyValueUpdatesInfo{};
+  auto category_hashers = std::map<std::string, Hasher>{};
+
+  for (auto it = update.kv.begin(); it != update.kv.end();) {
+    // Save the iterator as extract() invalidates it.
+    auto extracted_it = it;
+    ++it;
+    auto node = update.kv.extract(extracted_it);
+    auto &key = node.key();
+    auto &shared_value = node.mapped();
+    if (shared_value.category_ids.empty()) {
+      throw std::invalid_argument{"Empty category ID set for a shared key-value update"};
+    }
+
+    const auto versioned_key = versionedKey(key, block_id);
+
+    auto [key_it, inserted_key] = update_info.keys.emplace(std::move(key), SharedKeyData{});
+    // Make sure we've inserted the key for the first time.
+    ConcordAssert(inserted_key);
+    auto &key_data = key_it->second;
+
+    const auto value_hash = hash(shared_value.value);
+    auto first_category = std::optional<std::string>{std::nullopt};
+    for (auto &&category_id : shared_value.category_ids) {
+      const auto data_cf = dataCf(category_id);
+      const auto latest_version_cf = latestVersionCf(category_id);
+      createColumnFamilyIfNotExisting(data_cf);
+      createColumnFamilyIfNotExisting(latest_version_cf);
+
+      if (update.calculate_root_hash) {
+        updateCategoryHash(category_id, versioned_key.key_hash.value, value_hash, category_hashers);
+      }
+
+      // Versioned key-value.
+      if (!first_category) {
+        batch.put(data_cf, serialize(versioned_key), sharedValueData(std::move(shared_value.value)));
+        first_category = category_id;
+      } else {
+        batch.put(data_cf, serialize(versioned_key), sharedValuePointer(*first_category));
+      }
+
```

Comment:
An idea, maybe it would be simpler to have a dedicated column family for the value (global to the shared cat) where all other ```per category ID``` columns point to.
Then all there is to do is to mark versioned_key->True at the corresponding category ids.

Reply:
For example, how are you going to distinguish between a pointer or a value on get ?

Reply:
Currently, we serialize a variant and that's how we know which value is a pointer and which one is an actual value.

I think your solution is better, because we don't need to store the variant ID in the value and we don't need to parse it, we can just lookup in the category and find if the key is part of it or not.

Great catch, thanks! Will change it.

Reply:
After further discussions with @andrewjstone, key versions per category are implemented via the `KeyVersionsPerCategory` type that contains a mapping from a category ID to a list of key versions for that category ID. That allows implementations for `get()` methods without using RocksDB iterators. Assumption is that keys in the shared category will not have many versions. If we need many versions, we can define a new category.